README.md
<img height="180" width="100%" src="https://github.com/deltachat/deltaircd/raw/master/images/logo.svg" alt="deltaircd">

<p align="center">Minimal IRC server which integrates with <a href="https://delta.chat">Delta Chat</a></p>

## Features

- support direct messages / private channels
- auto-join/leave to same channels as on Delta Chat
- support multiple users (one Delta Chat accounts per IRC user/connection)
- support channel/direct message backlog (messages when you're disconnected from IRC/Delta Chat)
- WHOIS, WHO, JOIN, LEAVE, NICK, LIST, ISON, PRIVMSG, MODE, TOPIC, LUSERS, AWAY, KICK, INVITE support
- support TLS (ssl)
- support unix sockets
- &users channel that contains all your contacts for easy messaging
- support for including/excluding channels from showing up in IRC
- support multiline pasting
- support for adding account as second device (importing account from different device)
- search users (/msg deltachat searchusers query)
- search messages (/msg deltachat search query)
- scrollback support (/msg deltachat scrollback #channel limit)
- prefixcontext option (see <https://github.com/deltachat/deltaircd/blob/master/prefixcontext.md>)
  - send replies
  - send reactions
  - delete messages

## Installing dependencies

To use deltairc first make sure you have `deltachat-rpc-server` program installed in your
`PATH`, for more info check:
https://github.com/deltachat/deltachat-core-rust/tree/master/deltachat-rpc-server

## Building

Go 1.17+ is required

```bash
go install github.com/deltachat/deltaircd
```

You should now have `deltaircd` binary in the bin directory:

```bash
$ ls ~/go/bin/
deltaircd
```

## Config file

See [deltaircd.toml.example](https://github.com/deltachat/deltaircd/blob/master/deltaircd.toml.example)
Run with `deltaircd --conf deltaircd.toml`

## Usage

```
Usage of ./deltaircd:
      --bind string      interface:port to bind to, or a path to bind to a Unix socket. (default "127.0.0.1:6667")
      --conf string      config file (default "deltaircd.toml")
      --debug            enable debug logging
      --tlsbind string   interface:port to bind to. (e.g 127.0.0.1:6697)
      --tlsdir string    directory to look for key.pem and cert.pem. (default ".")
      --version          show version
```

deltaircd will listen by default on localhost port 6667.
Connect with your favorite IRC client to localhost:6667

For TLS support you'll need to generate certificates.
You can use this program [generate_cert.go](https://golang.org/src/crypto/tls/generate_cert.go) to
generate key.pem and cert.pem

### Delta Chat user commands

Configure a new account with email/pass and login into it

```
/msg deltachat login <email> <password>
```

Add account as second device (importing account from different device)

```
/msg deltachat login DCBACKUP:...
```

Login into existing previously configured accout

```
/msg deltachat login <email>
```

## Credits

deltaircd is a port of [matterircd](https://github.com/42wim/matterircd) for Delta Chat.


changelog.md
# Unreleased

- first release: adapted from matterircd


main.go
package main

import (
	"crypto/tls"
	"flag"
	"fmt"
	"log"
	"net"
	"os"
	"strings"

	"github.com/deltachat/deltaircd/config"
	irckit "github.com/deltachat/deltaircd/mm-go-irckit"
	"github.com/google/gops/agent"
	prefixed "github.com/matterbridge/logrus-prefixed-formatter"
	"github.com/sirupsen/logrus"
	"github.com/spf13/pflag"
	"github.com/spf13/viper"
)

var (
	version = "0.27.1-dev"
	githash string
	logger  *logrus.Entry
	v       *viper.Viper
)

func main() {
	ourlog := logrus.New()
	ourlog.Formatter = &prefixed.TextFormatter{
		PrefixPadding: 10,
		FullTimestamp: true,
	}
	logger = ourlog.WithFields(logrus.Fields{"prefix": "deltaircd"})
	config.Logger = logger

	// config related. instantiate a new config.Config to store flags
	flagConfig := flag.String("conf", "deltaircd.toml", "config file")

	// bools for showing version/enabling debug
	flag.Bool("version", false, "show version")
	flag.Bool("debug", false, "enable debug logging")

	// bind related cfg
	flag.String("bind", "127.0.0.1:6667", "interface:port to bind to, or a path to bind to a Unix socket.")

	// TLS related cfg
	flag.String("tlsbind", "", "interface:port to bind to. (e.g 127.0.0.1:6697)")
	flag.String("tlsdir", ".", "directory to look for key.pem and cert.pem.")

	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
	pflag.Parse()

	// Attempt to load values from the config file
	if _, err := os.Stat(*flagConfig); err == nil {
		v, err = config.LoadConfig(*flagConfig)
		if err != nil {
			log.Fatal(err)
		}
	} else {
		v = viper.New()
	}

	v.BindPFlags(pflag.CommandLine)

	if v.GetBool("debug") {
		logger.Info("enabling debug")
		ourlog.Level = logrus.DebugLevel
		irckit.SetLogLevel("debug")
	}

	if v.GetBool("trace") {
		logger.Info("enabling trace")
		ourlog.Level = logrus.TraceLevel
		irckit.SetLogLevel("trace")
	}

	if v.GetBool("gops") {
		if err := agent.Listen(agent.Options{}); err != nil {
			log.Fatal(err)
		}
	}

	if v.GetBool("version") {
		fmt.Printf("version: %s %s\n", version, githash)
		return
	}

	irckit.SetLogger(logger)

	logger.Infof("Running version %s %s", version, githash)
	if strings.Contains(version, "-dev") {
		logger.Infof("WARNING: THIS IS A DEVELOPMENT VERSION. Things may break.")
	}

	if v.GetString("tlsbind") != "" {
		go func() {
			logger.Infof("Listening on %s (TLS)", v.GetString("tlsbind"))
			socket := tlsbind()
			defer socket.Close()
			start(socket)
		}()
	}

	// backwards compatible

	if v.GetString("bind") != "" {
		go func() {
			var network string
			if strings.ContainsRune(v.GetString("bind"), os.PathSeparator) {
				network = "unix"
			} else {
				network = "tcp"
			}

			socket, err := net.Listen(network, v.GetString("bind"))
			if err != nil {
				logger.Errorf("Can not listen on %s: %v", v.GetString("bind"), err)
				os.Exit(1)
			}

			logger.Infof("Listening on %s", v.GetString("bind"))

			defer socket.Close()
			start(socket)
		}()
	}

	select {}
}

func tlsbind() net.Listener {
	certPath := v.GetString("tlsdir") + "/cert.pem"
	keyPath := v.GetString("tlsdir") + "/key.pem"

	if v.GetString("tlscert") != "" {
		certPath = v.GetString("tlscert")
	}

	if v.GetString("tlskey") != "" {
		keyPath = v.GetString("tlskey")
	}

	kpr, err := NewKeypairReloader(certPath, keyPath)
	if err != nil {
		logger.Errorf("could not load TLS, incorrect directory? Error: %s", err)
		os.Exit(1)
	}

	tlsConfig := tls.Config{
		GetCertificate: kpr.GetCertificateFunc(),
	}

	listenerTLS, err := tls.Listen("tcp", v.GetString("tlsbind"), &tlsConfig)
	if err != nil {
		logger.Errorf("Can not listen on %s: %v\n", v.GetString("tlsbind"), err)
		os.Exit(1)
	}

	logger.Info("TLS listening on ", v.GetString("tlsbind"))

	return listenerTLS
}

func start(socket net.Listener) {
	for {
		conn, err := socket.Accept()
		if err != nil {
			logger.Errorf("Failed to accept connection: %v", err)
			return
		}

		go func() {
			newsrv := irckit.ServerConfig{Name: "deltaircd", Version: version}.Server()

			logger.Infof("New connection: %s", conn.RemoteAddr())

			user := irckit.NewUserBridge(conn, newsrv, v)
			err = newsrv.Connect(user)
			if err != nil {
				logger.Errorf("Failed to join: %v", err)
				return
			}
		}()
	}
}
